cluster-default-publishing-strategy.md

```diff
@@ -0,0 +1,138 @@
+
+The field will be added to [`status` section][infrastructure-api-status] because the goal is to provide an API for discovery i.e. the current state, and not to provide an API for customers to specify the desired state.
+
+### User Stories
+
+#### Default Ingress Controller LoadBalancer Scope
+
+The ingress-operator will be able to use this field to define the **default** [scope][ingress-controller-api-load-balancer-scope] for the LoadBalancers for the [ingress controllers][ingress-controller-api].
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Addition of the new field requires migration of all the existing clusters. Since we have only supported `External` strategy for OpenShift, the migration should be simple enough.
+
+## Design Details
+
+### New API
+
+`PublishingStrategy` enum definition.
+
+```go
+// PublishingStrategy is a strategy for how various endpoints for the cluster are exposed.
+type PublishingStrategy string
+
+const (
+    // ExternalPublishingStrategy exposes endpoints for the cluster to the Internet
+    ExternalPublishingStrategy PublishingStrategy = “External”
+    // InternalPublishingStrategy exposes the endpoints for the cluster to the private network only.
+    InternalPublishingStrategy PublishingStrategy = “Internal”
+)
+```
+
+Updating the `InfrastructureStatus` object to add `Publish` field.
+
+```go
+type InfrastructureStatus struct {
+...
+
+// Publish controls how the user facing endpoints of the cluster like the k8s API, OpenShift routes etc. are exposed.
+// When no strategy is specified, the strategy is `External`.
+// +optional
+Publish PublishingStrategy `json:”publish,omitempty”`
+
+...
+
+}
+```
+
+### Migration of existing clusters
+
+All clusters with `empty` PublishingStrategy can be updated to `External`, while all non-empty values will be kept untouched.
+
+#### Rollbacks
+
+In case of rollbacks after the migration has been done, the value `External` will be accurate.
+
+### Test Plan
+
+#### Migration
+
+An e2e test for ensuring the migration is correctly performed during upgrades from 4.2 will be required.
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+See the [migration section][#migration-of-existing-clusters]
+
+### Version Skew Strategy
+
+The API defines `empty` equivalent to `External`, therefore all the clients will have to keep that check to remain backwards compatible.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+- Because the ingres-operator is currently the only user of this API field, the `PublishingStrategy` could also be added to the [`Ingress`][ingress-api] custom resource. But because the strategy affects more than the ingress, like the
+Network security filters on clouds, this field is more suited for `Infrastructure` custom resource in the long term.
```

Comment:
@ironcladlou claim here.

Reply:
The only use cases I know of so far related to this API would. be the installer-created API LB, and the ingress LB. Not sure about anything else.

Reply:
LBs require more than LB itself, it requires the network firewall rules are setup correctly. We need this information in API to make sure we can transfer a lot of the installer work into the bootstrap-host and controllers that manage these, because we don't today.